jgscm/__init__.py

```diff
@@ -522,13 +523,15 @@ def _get_bucket(self, name, throw=False):
             cache[name] = bucket
             return bucket
 
-    @staticmethod
-    def _parse_path(path):
+    def _parse_path(self, path):
         """
         Splits the path into bucket name and path inside the bucket.
         :param path: string to split.
         :return: tuple(bucket name, bucket path).
         """
+        if self.default_path and not path.startswith(self.default_path):
+            path = f"{self.default_path}/{path}"
```

Comment:
Ah, I think I'm starting to see: path is not a URL but it may or may not include the bucket.

Hmm, this doesn't do what the old code does.  If default_path doesn't have a leading slash, you probably need to add it in the call to `startswith`.  If path has a leading slash, then this code doesn't behave the same as the previous one.  You probably want something like `os.path.join(f'/{default_path}', path)`.

Reply:
default_path shouldn't have a leading slash https://github.com/src-d/jgscm#usage
  * `c.GoogleStorageContentManager.default_path = 'path/without/starting/slash'`

The _parse_path needs to return the bucket_name (no slash), and the rest of the path, with the slash separating bucket_name and the rest of the path stripped. The call right below my change is:

```py
 bucket, _, blobname = path.partition("/")
 return bucket, blobname
```

which splits on the first slash, and strips that first slash (because the middle argument is dropped).


The behavior we want:
* `_parse_path("your_bucket_name_with_slash/")` we should get `("your_bucket_name", "")`
* `_parse_path("some-folder")` we should get `("your_bucket_name", "some_folder")`
    * This would previously fail, you would get `("some_folder", "")`, which would result in the failed get operation.
* `_parse_path("some-folder/")` we should get `("your_bucket_name", "some_folder/")`
* `_parse_path("your_bucket_name/some-folder")` we should get `("your_bucket_name", "some_folder")`
* `_parse_path("your_bucket_name/some-folder/some_file")` we should get `("your_bucket_name", "some-folder/some_file")`

I've attached some examples of the call stack:

<img width="732" alt="Screenshot 2019-04-08 13 54 11" src="https://user-images.githubusercontent.com/5543229/55745752-f5ff5b00-5a05-11e9-9332-c8fd2946126e.png">
<img width="736" alt="Screenshot 2019-04-08 13 54 32" src="https://user-images.githubusercontent.com/5543229/55745753-f5ff5b00-5a05-11e9-800d-f6259e2beb0a.png">
<img width="834" alt="Screenshot 2019-04-08 13 54 52" src="https://user-images.githubusercontent.com/5543229/55745754-f5ff5b00-5a05-11e9-9882-905ade98bace.png">
<img width="878" alt="Screenshot 2019-04-08 13 55 14" src="https://user-images.githubusercontent.com/5543229/55745755-f5ff5b00-5a05-11e9-9d0f-238058963bab.png">

Reply:
The only behavior that this changes is that when you create a bucket in the root dir, and you have default_path set, it will use that default path, rather than treat the directory being created as a new bucket. That's why I had treated this as a breaking change.

And I realized it actually fixes another issue. In the old jgscm, when you created 2 python files in the root dir, the 2nd would overwrite the first. This fixes that as well.

Reply:
Enumerating the cases makes it clear this design has a real ambiguity when there is a folder with the same name as a bucket, but let's deal with that after this goes in.

Your code still breaks on the case there is a folder that starts with (but is not equal to) the bucket name, my recommendation of `startswith` is bad.

 > when you create a bucket in the root dir

You mean folder?

 > That's why I had treated this as a breaking change.

This really seems like a bug (creating things outside the default path), which is why I'd call it a patch release.

Reply:
> Enumerating the cases makes it clear this design has a real ambiguity when there is a folder with the same name as a bucket, but let's deal with that after this goes in.

Yes, I agree, I don't like the lack of separation between folder and bucket taken in jgscm. I think the more stable solution is to always prefix the filename with the bucket name, unless in the root folder, in which case the behavior will be different depending on the flag passed (and that maybe shouldn't be default_path; someone may want to be able to create buckets even when a default bucket is specified).

> Your code still breaks on the case there is a folder that starts with (but is not equal to) the bucket name, my recommendation of startswith is bad.

I will attempt to address this now.


> You mean folder?
Yes

Reply:
I believe it's addressed by checking for 

`path.startswith(f"{self.default_path}/")`

This wouldn't fix the case when the user tries to name the folder as the f"{bucket_name}/", but that seems wrong in any case.

Reply:
I propose that we defer the moving fix for another PR, so that Konrad/Liam can test other functions. However, can work further on this if you think worth it. Whatever you prefer :)

Reply:
I'd open an issue for the file/folder-as-bucket-name problem and make it low priority.  Getting IO (hadoop stuff) and user isolation going seem much more important.

Reply:
Done